pkg/templates/k8sresource_funcs.go

```diff
@@ -126,6 +126,38 @@ func (t *TemplateResolver) protect(value string) (string, error) {
 	return protectedPrefix + base64.StdEncoding.EncodeToString(encryptedValue), nil
 }
 
+// decrypt will decrypt a string that was encrypted using the protect method. An error is returned if the base64 or
+// the AES key is invalid.
+func (t *TemplateResolver) decrypt(value string) (string, error) {
+	decodedValue, err := base64.StdEncoding.DecodeString(value)
+	if err != nil {
+		return "", fmt.Errorf("%s: %w: %v", value, ErrInvalidB64OfEncrypted, err)
+	}
+
+	block, err := aes.NewCipher(t.config.AESKey)
+	if err != nil {
+		return "", fmt.Errorf("%w: %v", ErrInvalidAESKey, err)
+	}
+
+	// This is already validated in the NewResolver method, but is checked again in case that method was bypassed
+	// to avoid a panic.
+	if len(t.config.InitializationVector) != IVSize {
+		return "", ErrInvalidIV
+	}
+
+	blockMode := cipher.NewCBCDecrypter(block, t.config.InitializationVector)
+
```

Comment:
the above  code from line 137 to 148 -  does it generate a different decrypted everytime ?   does this need to be run  for decryption call or  can this block be executed in the template resolver/initializer and set on the templateResolver ?

Reply:
I think for line 137 we could move that up, but not line 148 because the [docs](https://pkg.go.dev/crypto/cipher#BlockMode) say ".. BlockMode maintains state and does not reset at each CryptBlocks call". With that in mind, I like abstracting the decryption code entirely in the `decrypt` method since it's a very small performance hit.